File: Other Versions/Go/utils/vscdb.go
package utils

import (
	"database/sql"
	"path/filepath"
	"sort"
	"strings"

	"aihound/core"

	// Pure-Go SQLite driver (no CGO required).
	// Dependency: modernc.org/sqlite — add to go.mod with:
	//   go get modernc.org/sqlite
	_ "modernc.org/sqlite"
)

// GetVSCDBPaths returns the possible paths to VS Code's state.vscdb file
// for the current platform.
func GetVSCDBPaths() []string {
	plat := core.DetectPlatform()
	var paths []string

	switch plat {
	case core.PlatformLinux, core.PlatformWSL:
		cfg := core.GetXDGConfig()
		if cfg != "" {
			paths = append(paths, filepath.Join(cfg, "Code", "User", "globalStorage", "state.vscdb"))
		}
	case core.PlatformMacOS:
		home := core.GetHome()
		if home != "" {
			paths = append(paths, filepath.Join(home, "Library", "Application Support", "Code", "User", "globalStorage", "state.vscdb"))
		}
	case core.PlatformWindows:
		appdata := core.GetAppData()
		if appdata != "" {
			paths = append(paths, filepath.Join(appdata, "Code", "User", "globalStorage", "state.vscdb"))
		}
	}

	// WSL also checks the Windows-side AppData path.
	if plat == core.PlatformWSL {
		appdata := core.GetAppData()
		if appdata != "" {
			paths = append(paths, filepath.Join(appdata, "Code", "User", "globalStorage", "state.vscdb"))
		}
	}

	return paths
}

// ListSecretKeys opens the VS Code state.vscdb SQLite database and returns
// all keys from the ItemTable that begin with "secret://".
// The values are encrypted with Electron's safeStorage API and cannot be
// decrypted without the OS keychain encryption key.
func ListSecretKeys(vscdbPath string) ([]string, error) {
	db, err := sql.Open("sqlite", vscdbPath)
	if err != nil {
		return nil, err
	}
	defer db.Close()

	rows, err := db.Query("SELECT key FROM ItemTable WHERE key LIKE 'secret://%'")
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var keys []string
	for rows.Next() {
		var key string
		if err := rows.Scan(&key); err != nil {
			continue
		}
		keys = append(keys, key)
	}
	return keys, rows.Err()
}

// GetExtensionIDsWithSecrets returns sorted, unique extension IDs that have
// stored secrets in the given VS Code state.vscdb file.
// Secret keys have the format: secret://<extension-id>/<secret-name>
func GetExtensionIDsWithSecrets(vscdbPath string) ([]string, error) {
	keys, err := ListSecretKeys(vscdbPath)
	if err != nil {
		return nil, err
	}

	seen := make(map[string]struct{})
	for _, key := range keys {
		if !strings.HasPrefix(key, "secret://") {
			continue
		}
		rest := key[len("secret://"):]
		parts := strings.SplitN(rest, "/", 2)
		if len(parts) > 0 && parts[0] != "" {
			seen[parts[0]] = struct{}{}
		}
	}

	extensions := make([]string, 0, len(seen))
	for ext := range seen {
		extensions = append(extensions, ext)
	}
	sort.Strings(extensions)
	return extensions, nil
}

File: aihound/utils/vscdb.py
"""VS Code state.vscdb SQLite reader.

VS Code stores extension secrets in a SQLite database (state.vscdb)
encrypted with Electron's safeStorage API (AES-128-CBC).
We can detect the presence of stored secrets but cannot decrypt them
without the OS keychain encryption key.
"""

from __future__ import annotations

import sqlite3
from pathlib import Path
from typing import Optional

from aihound.core.platform import (
    detect_platform, Platform, get_home, get_appdata, get_xdg_config,
)


def get_vscdb_paths() -> list[Path]:
    """Get possible paths to VS Code's state.vscdb file."""
    plat = detect_platform()
    paths = []

    if plat in (Platform.LINUX, Platform.WSL):
        paths.append(get_xdg_config() / "Code" / "User" / "globalStorage" / "state.vscdb")
    elif plat == Platform.MACOS:
        paths.append(
            get_home() / "Library" / "Application Support" / "Code"
            / "User" / "globalStorage" / "state.vscdb"
        )
    elif plat == Platform.WINDOWS:
        appdata = get_appdata()
        if appdata:
            paths.append(appdata / "Code" / "User" / "globalStorage" / "state.vscdb")

    if plat == Platform.WSL:
        appdata = get_appdata()
        if appdata:
            paths.append(appdata / "Code" / "User" / "globalStorage" / "state.vscdb")

    return paths


def list_secret_keys(vscdb_path: Path) -> list[str]:
    """List keys in the VS Code secret storage.

    Returns key names that have stored secrets (values are encrypted).
    """
    if not vscdb_path.exists():
        return []

    try:
        conn = sqlite3.connect(str(vscdb_path))
        cursor = conn.cursor()

        # The secrets are stored in the ItemTable with keys like:
        # secret://<publisher>.<extension>/<secret-name>
        cursor.execute(
            "SELECT key FROM ItemTable WHERE key LIKE 'secret://%'"
        )
        keys = [row[0] for row in cursor.fetchall()]
        conn.close()
        return keys

    except (sqlite3.Error, OSError):
        return []


def get_extension_ids_with_secrets(vscdb_path: Path) -> list[str]:
    """Get unique extension IDs that have stored secrets."""
    keys = list_secret_keys(vscdb_path)
    extensions = set()
    for key in keys:
        # Format: secret://<extension-id>/<secret-name>
        if key.startswith("secret://"):
            parts = key[len("secret://"):].split("/", 1)
            if parts:
                extensions.add(parts[0])
    return sorted(extensions)
